Tests/NIOTests/EventLoopTest.swift

```diff
@@ -279,4 +279,16 @@ public class EventLoopTest : XCTestCase {
         _ = eventLoop.scheduleTask(in: .hours(1)) { }
         try group.syncShutdownGracefully()
     }
+
+    public func testCloseFutureNotifiedBeforeUnblock() throws {
+        let group = MultiThreadedEventLoopGroup(numThreads: 1)
+        let eventLoop = group.next()
+
+        let channel = try SocketChannel(eventLoop: eventLoop as! SelectableEventLoop, protocolFamily: AF_INET)
+        try channel.register().wait()
+        XCTAssertFalse(channel.closeFuture.isFulfilled)
+        try group.syncShutdownGracefully()
+        XCTAssertTrue(channel.closeFuture.isFulfilled)
+        XCTAssertFalse(channel.isActive)
```

Comment:
This test would probabilistically pass before the given change: it relies on a race between `syncShutdownGracefully` returning and the check on the state of `channel.closeFuture and channel.isActive`. These checks should probably be done by inserting a channel handler that catches `close` and checks whether `syncShutdownGracefully` has returned yet. It may even want to wait a few ms to be sure.

Reply:
@Lukasa good idea... I think we should add the check in channelInactive(...) tho. WDYT ?

Reply:
Actually `handlerRemoved(...)`